### PR TITLE
Add code formatting and linting

### PR DESCRIPTION
**Pull request by danpat**
# Issue

This is a replacement PR for https://github.com/valhalla/valhalla/pull/1250.

This PR:

  1. Formats all code using `clang-format`
  2. Adds a new CI build that checks that code adheres to expected format (and fails if it doesn't)
  3. Adds some minimum `clang-tidy` checks that also fail the build if they don't pass.

This is a *large* PR that touches many files across the codebase.  I've broken the commits down into what I think are reasonable sized pieces that should look OK when reviewing code history.

## Tasklist

 - [ ] review - you must request approval to merge any PR to master

### CODE REVIEW

File: valhalla/sif/edgelabel.h [hunk 1]
```diff
@@ -364,20 +359,20 @@ class EdgeLabel {
...
-  uint64_t origin_           : 1;
-  uint64_t toll_             : 1;
-  uint64_t not_thru_         : 1;
-  uint64_t deadend_          : 1;
-  uint64_t on_complex_rest_  : 1;
-
-  Cost cost_;       // Cost and elapsed time along the path.
-  float sortcost_;  // Sort cost - includes A* heuristic.
-  float distance_;  // Distance to the destination.
+  uint64_t endnode_ : 48; // Could be 46 (2 spare)
```
Comment (dnesbitt61) [inline valhalla/sif/edgelabel.h]: I like the way the bitfield sizes were lined up so they are easy to add. Guess I can get over this change

Comment (danpat) [reply valhalla/sif/edgelabel.h]: Hmm, I agree this is unfortunate.  There doesn't seem to be a styling rule available to configure this.

There's one ticket that's sort of related: https://bugs.llvm.org/show_bug.cgi?id=19067

It makes the point that one problem that arises with this kind of alignment is that introduction of a new long variable in the list can change minimum column width, and require re-formatting all variables in the block to re-align them, adding formatting noise to each diff unrelated to the change.

@dnesbitt61 We could disable formatting on these sections to keep the alignment - I do kind of like it to.

http://clang.llvm.org/docs/ClangFormatStyleOptions.html#disabling-formatting-on-a-piece-of-code

Do you feel strongly enough about it for me to update these blocks?

Comment (dnesbitt61) [reply valhalla/sif/edgelabel.h]: No. actually the comment about a long name requiring a whole block to be changed to re-align makes me favor the new way. I keep forgetting how many times I have had to realign things. Of course it sort of makes me feel good to "keep order" in the code but it definitely wastes time.

Comment (danpat) [reply valhalla/sif/edgelabel.h]: 👍 I think we're ready to merge then!

**Pull request merged**